### ISSUE TITLE
[Desktop] Google components automatic update fails

### ISSUE THREAD

**Issue by jumde** [brave/brave-browser#10464]
## Description 
Automatic updates for google components fail.

## Steps to Reproduce

   1. Navigate to brave://components
   2. Wait for a few minutes

## Actual result:
Components like CRLSets error out

## Expected result:
No error

## Reproduces how often: 
Easily

## Brave version (brave://version info)
1.12.42 Chromium: 83.0.4103.106 (Official Build) nightly (64-bit)

**Comment by mherrmann** [comment]
Compiling some findings here.

Starting Brave Nightly with flags `--enable-logging=stderr --v=1` on Linux produces the following output when the update errors happen:

```
[21534:21534:1110/101526.834284:VERBOSE1:component_updater_service.cc(324)] CheckForUpdates: automatic updatecheck for components.
[21563:21638:1110/101526.841124:VERBOSE1:network_delegate.cc(32)] NetworkDelegate::NotifyBeforeURLRequest: https://go-updater.brave.com/extensions
[21534:21534:1110/101527.138161:VERBOSE1:request_sender.cc(181)] Request completed from url: https://update.googleapis.com/service/update2/json
[21534:21534:1110/101527.143297:VERBOSE1:component_updater_service.cc(411)] Update completed with error 0
```

I don't yet know whether Brave really connects to https://update.googleapis.com/service/update2/json, or whether that URL is actually rewritten to a Brave URL.

**Comment by mherrmann** [comment]
@mihaiplesa do you know who maintains https://go-updater.brave.com/extensions? I have a suspicion that the problem described in this issue is caused by a bug in that server implementation.

**Comment by mihaiplesa** [comment]
@jumde @linhkikuchi might be able to point in the right direction.

**Comment by mherrmann** [comment]
Hi @jumde @linhkikuchi I think https://go-updater.brave.com/extensions may be returning invalid responses in some cases. Specifically: Brave's component update check sends a request to this URL asking "I have components 1,2,3, is there an update?". But the URL only responds "there is no update for component 2". It should, I think, respond "there are no updates for components 1, 2, 3". Google's implementation https://update.googleapis.com/service/update2/json does this. Further details are above. Are you able to look into this, do you know who can look into it, or can you give me instructions how I can look into it?

**Comment by jumde** [comment]
@mherrmann - Code for go-updater.brave.com is here, `https://github.com/brave/go-update`. For google-components, go-updater redirect to `componentupdater.brave.com` which is a proxy to `update.googleapis.com`.

**Comment by mherrmann** [comment]
Thanks @jumde!

**Comment by mherrmann** [comment]
@jumde I looked at go-update and it seems to confirm my suspicion. What do you think of my hypothesis? Here it is again, in detail: I suspect that the update _client_ expects to receive a `noupdate` response for all components it requests an update for. But go-update only returns `noupdate` for those components which are in DynamoDb. This is unlike Google's implementation, https://update.googleapis.com/service/update2/json, which gives `noupdate`s for all components.

If you think my hypothesis is correct: Should I submit a PR to go-update that fetches and caches responses from componentupdater.brave.com?

**Comment by mherrmann** [comment]
@jumde an update please? I'm blocked on your response.

**Comment by jumde** [comment]
@mherrmann - If the component is not in the `DynamoDB`, the request is redirected to `https://componentupdater.brave.com` which is a proxy to `https://update.googleapis.com`. So the response should be `noupdate`. 

Let me dig into this a bit more to see what's going on, will share an update shortly.

**Comment by mherrmann** [comment]
@jumde advised me via Slack how this can be fixed. He agrees with my analysis of the cause of the problem. However, my [proposed solution](https://github.com/brave/brave-browser/issues/10464#issuecomment-725326012) won't work. For legal reasons, the Widevine component must be fetched from Google's servers. We are not allowed to set up a Brave proxy that mirrors Google's updates to this component.

Just to recapitulate: In an automatic update check, the component updater asks "are there updates for <brave component 1>, <brave component 2>, ..., <google component 1>, <google component 2>, ...". For legal reasons, Brave's update server can only return responses for Brave's components. But the updater expects to also receive a response for the other components it requested. This leads to the error described in this issue.

An important detail is that Brave's update server (go-updater) _is_ able to handle update requests for a single _Google_ component. The reason for this is that in this case it returns a redirect to Google's update server. This is what happens in on-demand updates of single components, hence why those updates do work.

To fix the problem, @jumde suggested we iterate over the components one by one. I'm looking into how best to implement this. Currently, I'm torn between [`component_updater_service`](https://chromium.googlesource.com/chromium/src/+/master/components/component_updater/component_updater_service.cc#319), [`update_client`](https://chromium.googlesource.com/chromium/src/+/master/components/update_client/update_client.cc#109) or even deeper.

**Comment by mherrmann** [comment]
Verified that my fix works in [Nightly v1.20.65](https://github.com/brave/brave-browser/releases/tag/v1.20.65) on Linux and Android.

Linux:

![image](https://user-images.githubusercontent.com/1076393/103923143-08478500-5115-11eb-8373-b3216f7d84c3.png)

Android:

![image](https://user-images.githubusercontent.com/1076393/103923213-20b79f80-5115-11eb-8654-9ab978f63fd5.png)

**Comment by GeetaSarvadnya** [comment]
Verification passed on

```

Brave | 1.20.86 Chromium: 88.0.4324.96 (Official Build) dev (64-bit)
-- | --
Revision | 68dba2d8a0b149a1d3afac56fa74648032bcf46b-refs/branch-heads/4324@{#1784}
OS | Windows 10 OS Version 2004 (Build 19041.746)

```
- Verified STR from the description

<details>
<summary>Clean profile</summary>

Confirmed `CRLSet` is listed on brave://components as `6376 component updated` on a clean profile.
Confirmed after ~1-minute reload component page is showing the component as `up-to-date` and same version is reflected.

Example | Example
----- | -----
![image](https://user-images.githubusercontent.com/38657976/105374983-4a50ea80-5c2e-11eb-95f8-4f2d69dcb258.png) | ![image](https://user-images.githubusercontent.com/38657976/105374868-28effe80-5c2e-11eb-9f0c-f4df78c28970.png)

</details>

<details>
<summary>Upgraded profile</summary>

Installed 1.19.78 (Dev) and confirmed `CRLSet` is listed on brave://components as `6376 Updated`.
After ~1 minute component showed Update Error.
Updated to 1.20.86 (Dev).
Confirmed `CRLSet` is listed on brave://components as `6376 Component not updated`.
Confirmed after ~1 minute reload component page is showing the component as `up-to-date` and same version is reflected.

Example 1.19.x | Example 1.19.x | Example 1.20.x | Example 1.20.x
----- | ----- | ----- | -----

![image](https://user-images.githubusercontent.com/38657976/105377852-3eb2f300-5c31-11eb-985c-13cea44229a6.png) | ![image](https://user-images.githubusercontent.com/38657976/105377027-6190d780-5c30-11eb-9a36-58d6618900fd.png) | ![image](https://user-images.githubusercontent.com/38657976/105377181-8a18d180-5c30-11eb-9deb-5367504e2ccd.png) | ![image](https://user-images.githubusercontent.com/38657976/105377451-d532e480-5c30-11eb-9b6c-3ae68ccc5338.png)



</details>
